pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,218 @@
+		row, err := d.InternalExecutor.QueryRow(ctx, "get_database_desc", txn, `
+SELECT id, descriptor, crdb_internal_mvcc_timestamp FROM system.descriptor WHERE id = $1`, dbID)
+		if err != nil {
+			return err
+		}
+		_, desc, _, err := unmarshalDescFromDescriptorRow(row)
+		if err != nil {
+			return err
+		}
+		if err != nil {
+			return err
+		}
+
+		mutableDesc := dbdesc.NewBuilder(desc.GetDatabase())
+		dbDesc := mutableDesc.BuildExistingMutable()
+
+		publicSchemaID, err := catalogkv.GenerateUniqueDescID(ctx, d.DB, d.Codec)
+		if err != nil {
+			return err
+		}
+		// Every database must be initialized with the public schema.
+		// Create the SchemaDescriptor.
+		// In postgres, the user "postgres" is the owner of the public schema in a
+		// newly created db. Postgres and Public have USAGE and CREATE privileges.
+		// In CockroachDB, root is our substitute for the postgres user.
+		publicSchemaPrivileges := descpb.NewDefaultPrivilegeDescriptor(security.AdminRoleName())
+		// By default, everyone has USAGE and CREATE on the public schema.
+		publicSchemaPrivileges.Grant(security.PublicRoleName(), privilege.List{privilege.CREATE, privilege.USAGE})
+		publicSchemaDesc := schemadesc.NewBuilder(&descpb.SchemaDescriptor{
+			ParentID:   dbID,
+			Name:       "public",
+			ID:         publicSchemaID,
+			Privileges: publicSchemaPrivileges,
+			Version:    1,
+		}).BuildCreatedMutableSchema()
+
+		descID := publicSchemaDesc.GetID()
+		idKey := catalogkeys.MakeSchemaNameKey(d.Codec, dbID, publicSchemaDesc.GetName())
+
+		key := catalogkeys.EncodeNameKey(d.Codec, catalogkeys.NewNameKeyComponents(dbID, keys.RootNamespaceID, tree.PublicSchema))
+
+		b := txn.NewBatch()
+
+		// Remove namespace entry for old public schema.
+		b.Del(key)
+		b.CPut(idKey, descID, nil)
+		if err := catalogkv.WriteNewDescToBatch(
+			ctx,
+			false,
+			d.Settings,
+			b,
+			d.Codec,
+			descID,
+			publicSchemaDesc,
+		); err != nil {
+			return err
+		}
+
+		dbDesc.DescriptorProto().GetDatabase().Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
+			tree.PublicSchema: {
+				ID: publicSchemaID,
+			},
+		}
+		if err := catalogkv.WriteDescToBatch(
+			ctx, false, d.Settings, b, d.Codec, dbDesc.GetID(), dbDesc,
+		); err != nil {
+			return err
+		}
+
+		if err := migrateObjectsInDatabase(ctx, dbID, d, txn, b, publicSchemaID); err != nil {
+			return err
+		}
+
+		return txn.Run(ctx, b)
+	})
+}
+
+func migrateObjectsInDatabase(
+	ctx context.Context,
+	dbID descpb.ID,
+	d migration.TenantDeps,
+	txn *kv.Txn,
+	batch *kv.Batch,
+	newPublicSchemaID descpb.ID,
+) error {
+	// Get All Descriptors under the current database's public schema.
+	rows, err := d.InternalExecutor.QueryIterator(ctx, "get_objects_in_db_public_schema", txn,
+		`SELECT d.id, descriptor, d.crdb_internal_mvcc_timestamp FROM system.descriptor d JOIN system.namespace ns ON d.id = ns.id WHERE ns."parentID" = $1
+AND ns."parentSchemaID" = 29`, dbID)
+	if err != nil {
+		return err
+	}
```

Comment:
We have some things to retrieve batches of descriptors from IDs and we have some things to get the set of descriptors in a schema. Can we use those instead of this? We don't currently have good batched retrieval in the descs.Collection, and that does make me sad, I'm inclined to ask you to write this using the highest level catalog APIs and then it'll motivate me or somebody to work on batched retrieval.

Reply:
Added a function `GetAllDescriptorsInDatabase`. I think this is where we want to add the batching right? I guess we could turn this `GetAllDescriptorsInDatabase` function into an iterator with batch sizes if we're worried about memory.